server/application/application.go

```diff
@@ -345,6 +345,31 @@ func (s *Server) UpdateSpec(ctx context.Context, q *ApplicationUpdateSpecRequest
 	return nil, status.Errorf(codes.Internal, "Failed to update application spec. Too many conflicts")
 }
 
+// PatchSpec patches an application spec
+func (s *Server) PatchSpec(ctx context.Context, q *ApplicationPatchSpecRequest) (*appv1.Application, error) {
+
+	app, err := s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Get(*q.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if !s.enf.Enforce(ctx.Value("claims"), rbacpolicy.ResourceApplications, rbacpolicy.ActionUpdate, appRBACName(*app)) {
+		return nil, grpc.ErrPermissionDenied
+	}
+
+	app, err = s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Patch(app.Name, types.PatchType(q.PatchType), []byte(q.Patch))
```

Comment:
We need to validate before patching. You can apply patch in memory using `jsonpatch` library. See example here: https://github.com/argoproj/argo-cd/blob/master/util/argo/diff_normalizer.go#L67